documentation/src/main/asciidoc/cassandraclient.adoc

```diff
@@ -0,0 +1,347 @@
+[source, java]
+----
+import javax.enterprise.inject.Produces;
+
+public class FruitMapperProducer {
+  private final CqlSession cqlSession;
+
+  @Inject
+  public FruitMapperProducer(CqlSession cqlSession) {
+    this.cqlSession = cqlSession;
+  }
+
+  @Produces
+  @ApplicationScoped
+  FruitMapper produceFruitMapper() {
+    return new FruitMapperBuilder(cqlSession).build();
+  }
+}
+----
+Please note the `CqlSession` instance injected automatically by the cassandra-quarkus extension in the `FruitMapperProducer` constructor.
+The `produceFruitMapper()` method is constructing `FruitMapper` that will be injected into the Di container by the Quarkus.
+
+Now create a `com.datastax.oss.quarkus.FruitService` that will be the business layer of our application and store/load the fruits from the Cassandra database.
+[source,java]
+----
+@ApplicationScoped
+public class FruitService {
+  private final FruitDao dao;
+
+  @Inject
+  public FruitService(FruitMapper fruitMapper, FruitServiceConfig fruitServiceConfig) {
+    dao = fruitMapper.fruitDao(CqlIdentifier.fromCql(fruitServiceConfig.keyspace));
+  }
+
+  public void save(Fruit fruit) {
+    dao.update(fruit);
+  }
+
+  public List<Fruit> get(String id) {
+    return dao.findById(id).all();
+  }
+}
+----
+Please note the construction of `FruitDao` instance in the constructor. It uses a `FruitMapper` instance injected automatically by the `FruitMapperProducer`.
+It is also using the injected `FruitServiceConfig` with the keyspace provided via the `application.properties` file.
+
+The last missing piece is the REST API that will expose GET and POST methods:
+[source,java]
+----
+@Path("/fruits")
+@Produces(MediaType.APPLICATION_JSON)
+@Consumes(MediaType.APPLICATION_JSON)
+public class FruitResource {
+
+  private static final String STORE_NAME = "acme";
+  @Inject FruitService fruitService;
+
+  @GET
+  public List<Fruit> list() {
+    return fruitService.get(STORE_NAME);
+  }
+
+  @POST
+  public void add(FruitDto fruit) {
+    fruitService.save(covertFromDto(fruit));
+  }
+
+  private Fruit covertFromDto(FruitDto fruitDto) {
+    return new Fruit(fruitDto.getName(), fruitDto.getDescription(), STORE_NAME);
+  }
+}
+----
+
+The `list` and `add` operations are executed for the `STORE_NAME` id. This is a partition key of the data model.
+We can easily retrieve all rows from cassandra using the partition key. They will be sorted by the clustering key.
+`FruitResource` is using `FruitService` which encapsulates the data access logic.
+
+When creating the REST API we should not share the same entity object between REST API and data access layers.
+They should not be coupled to allow the API to evolve independently of the storage layer.
+This is the reason why the API is using a `FruitDto` class.
+This class will be used by Quarkus to convert JSON to java object for client request and java object to JSON for the response.
```

Comment:
I agree that this is the right decision architecturally... but it's not what we're doing right now.

More specifically we are doing the right thing in the POST handler; we're expecting a JSON rep of a FruitDto and working with it as such.  But the handler for the GET makes no use of FruitDto at all.  And there's nothing in Quarkus or REST Easy that knows that we should render Fruit as a FruitDto instead (or knows how to translate one into the other).

To validate, note what we get with a simple curl request to the Web service:

```
[mersault@localhost ~]$ curl http://localhost:8080/fruits
[{"description":"foo","id":"acme","name":"bar"},{"description":"bar","id":"acme","name":"foo"}]
```

The "id" field is not in FruitDto... but it is in Fruit.  What's actually happening here is that the collection of Fruit objects returned from FruitResource.list() is being serialized to JSON directly.  But, as the comment suggests, we should be using FruitDto for this operation.

We need something like the following instead:

```java
  @GET
  public List<FruitDto> list() {
    return fruitService.get(STORE_NAME).stream()
            .map(fruit -> new FruitDto(fruit.getName(), fruit.getDescription()))
            .collect(Collectors.toList());
  }
```

Note the output of the curl request with this change:

```
[mersault@localhost ~]$ curl http://localhost:8080/fruits
[{"description":"foo","name":"bar"},{"description":"bar","name":"foo"}]
```